### PR TITLE
fix: harden notify chaining repair edge cases

### PR DESCRIPTION
**Pull request by lamther**
## Summary
- fail closed when original notify commands are wrapped through node/bun/deno without an explicit readable script target
- normalize /usr/bin/env wrappers more conservatively, including -S split strings, -- assignments, value-taking env options, and unknown env options
- preserve valid explicit-script chaining, restore no-backup-not-installed reporting, and queue a follow-up popover refresh when a load is already in flight

## Verification
- node --test test/init-uninstall.test.js test/codex-config-multiline-notify.test.js test/macos-popover-anchor-window.test.js
- npm test
- git diff --check && npm run validate:guardrails && node --test test/architecture-guardrails.test.js
- xcodebuild build -scheme TokenTrackerBar -destination 'platform=macOS'

Context: #291 was merged before the final notify-wrapper edge-case fixes landed on the fork branch, so this follow-up is based on the latest main and contains only the remaining hardening patch.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved notification command handling on Windows, including support for `~\` paths.
  - Prevented chaining to missing, unreadable, or incomplete interpreter-based notification scripts.
  - Hardened parsing and updating of multiline TOML notification settings.
  - Preserved unrelated and table-scoped configuration during notification updates and restores.
  - Improved recovery when notification settings are malformed or backups are unavailable.
- **Tests**
  - Expanded coverage for malformed TOML, interpreter options, Windows paths, and runtime-specific notification commands.

### CODE REVIEW

**Review by xiufengsun — CHANGES_REQUESTED** [review]
## P1 issues found during review

### 1. Double `loadAll()` doubles the menu-bar wait time
`DashboardViewModel.swift:282-284` — when `isLoading` is true, the code enqueues a **second** `loadAll()` before checking `isSyncing`. On first launch or wake-sync, this causes two full load rounds to run back-to-back.

Real-world impact: a lightweight sync completes in ~0.22 s, but `usage-limits` takes ~10.5 s. Two sequential rounds push the visible wait from ~10 s to ~20 s.

### 2. TOML multi-line string repair leaves a corrupted config
`src/lib/codex-config.js:374-420` — the unterminated multi-line string state carries across lines. `upsertNotify()` replaces the first line, but the **remaining lines of the old array literal stay in the file**.

Repro result:
```toml
notify = ["/usr/bin/env", "node", "/tmp/managed.cjs"]
  "/usr/bin/env",
  "unterminated
  "/tmp/hook.js"
]
```
Codex still cannot parse this config.

### 3. `env --chdir` notifications are silently dropped
`src/commands/init.js:1213-1243,1692-1700` — the parser strips `--chdir` but then checks the **relative** script path against TokenTracker's own cwd instead of the chdir target. A legitimate command like:
```
env --chdir /path/to/hooks node ./hook.js
```
is flagged as unreadable and silently skipped. Targeted repro confirms `marker=missing`.

**Comment by lamther** [comment]
Thanks for the detailed review. I pushed an updated commit: `0e14d5c0d9fa2d67616b21e89e9604f2d9d0633a`.

What changed:

1. Dashboard queued reloads
- `loadAll()` no longer drops refresh requests that arrive during an already queued reload pass.
- The queued reload drain now loops until the pending flag is quiet, so A in-flight / B queued / C during B results in one final follow-up instead of losing C.
- I also kept `isLoading = true` through `WidgetSnapshotWriter.update(...)` and made the drain guard require `!isLoading`, which closes the reentrant window where another load could start during widget snapshot writing.

2. Codex notify TOML repair
- `read/set/remove` now operate only on the top-level `notify` before the first TOML table.
- Table-scoped `notify` entries are preserved.
- Valid single-line arrays are distinguished from damaged/unterminated blocks, so following dotted keys, quoted keys, comments, and EOF comments are not consumed by damaged-block cleanup.

3. `/usr/bin/env` and cwd-aware command validation
- `env --chdir`, `--chdir=`, and `-C` now feed an effective cwd into explicit relative command validation.
- Env-wrapped explicit generic relative commands now fail closed when missing under that cwd.
- Bun `--cwd` and `--cwd=` now also define the base directory for relative script validation, so cwd-relative scripts are accepted and process-cwd false positives are rejected.

I added regression coverage for the specific edges above:
- top-level vs table-scoped notify
- single-line notify followed by dotted/quoted/comment boundaries
- restore/remove preserving table-scoped notify
- env `--chdir` and `-C` generic missing relative commands
- Bun `--cwd` and `--cwd=` allow/reject cases
- Dashboard static guards for queued drain and the reentrant loading window

Local verification after the final commit:
- `node --test test/init-uninstall.test.js test/codex-config-multiline-notify.test.js test/macos-popover-anchor-window.test.js` -> 50/50 pass
- `npm test` -> 1352/1352 pass
- `git diff --check` -> pass
- `npm run validate:guardrails` -> pass
- `xcodebuild build -scheme TokenTrackerBar -destination 'platform=macOS'` -> BUILD SUCCEEDED
- Final read-only cross-check on the final worktree converged: 4/4 votes, gaps=0.

**Review by xiufengsun — CHANGES_REQUESTED** [review]
## Follow-up review — 1 remaining P1

Thanks for fixing the TOML corruption and `env --chdir` issues — both look solid now.

### [P1] Popover still queues a redundant full reload

`DashboardViewModel.swift:291-293`:
```swift
if isLoading {
    queueReloadAfterCurrentLoad()
    return
}
```

When a `loadAll()` is already in flight, opening the popover sets the queue flag. Once the current load finishes, `runQueuedReloadIfNeeded()` unconditionally fires **another** full `loadAll()` — 9 API calls + Limits + widget snapshot, all over again.

The `!isSyncing` guard is still placed **after** the queuing branch, so a sync-triggered load will also accept the queued reload.

v0.74.4's cached-data display masks the visual stutter, but the duplicate network round-trip is still there.

**Suggested fix:** guard early in `refreshForPopoverOpen()`:

```swift
guard !isLoading, !isSyncing else { return }
```

The queued-reload mechanism can stay for scenarios that genuinely need a final-state refresh (e.g. period switching), but popover open should not trigger it.

`test/macos-popover-anchor-window.test.js:100-108` also needs updating — it currently locks in the queuing behavior as expected.

---

Everything else checks out: CI, CodeQL, CodeRabbit all green; no merge conflicts with main; TOML and chdir fixes verified. This is the last blocker.

**Comment by lamther** [comment]
Thanks, agreed on this distinction. I pushed the follow-up fix in `0725dcbdebadffe483d2eb9029114c2a5f003ed8`.

What changed:

- `refreshForPopoverOpen()` is now an opportunistic freshness check: it returns early with `guard !isLoading, !isSyncing else { return }`, so opening the popover cannot enqueue a second full `loadAll()` behind an already-running slow load or sync-triggered load.
- The queued-reload mechanism stays in `loadAll()` for explicit/internal reload requests that genuinely need a final-state refresh. That path still coalesces concurrent requests and drains until quiet, and `isLoading` remains true through the widget snapshot write.
- `test/macos-popover-anchor-window.test.js` now locks in the corrected split: `refreshForPopoverOpen()` must not call `queueReloadAfterCurrentLoad()`, while `loadAll()` still must queue one follow-up for concurrent explicit reloads.
- I also tightened that static regression test to extract the individual Swift function bodies before matching, so the `loadAll` and popover-open assertions cannot pass because of neighboring function text.

Validation on the final commit:

- `node --test test/init-uninstall.test.js test/codex-config-multiline-notify.test.js test/macos-popover-anchor-window.test.js` -> 50/50 pass
- `npm test` -> 1352/1352 pass
- `npm run validate:guardrails` -> pass
- `git diff --check` -> pass
- `xcodebuild build -scheme TokenTrackerBar -destination 'platform=macOS'` -> BUILD SUCCEEDED
- Final read-only cross-check on the final worktree converged: 4/4 votes, gaps=0.

**Comment by mm7894215** [comment]
Re-reviewed properly. The notify chaining hardening here is solid and I want to merge it — I confirmed it cherry-picks cleanly onto current main, is compatible with the recently-merged #312, and the notify/init tests pass (55/0).

Could you rebase onto the latest main and trim the PR to just the notify hardening? Specifically:

- Keep: `src/commands/init.js`, `src/lib/codex-config.js`, `test/codex-config-multiline-notify.test.js`, `test/init-uninstall.test.js`
- Drop: the popover-reload change (`DashboardViewModel.swift` + `test/macos-popover-anchor-window.test.js`) — main already has an equivalent guard (`refreshForPopoverOpen` skips while a load/sync is in flight via `syncInFlight`, `loadAll` queues one follow-up via `shouldReloadAfterCurrentLoad`), so that part now conflicts and is redundant.

Once it's rebased and slimmed down, CI will run and I'll merge. Thanks for the hardening work.

**Comment by lamther** [comment]
Thanks for the precise scope. I rebased the branch onto current `main` (`4b8a9ef`) and force-updated it to `ee2e870`.

The PR is now a single notify-hardening commit and changes only the four requested files:

- `src/commands/init.js`
- `src/lib/codex-config.js`
- `test/codex-config-multiline-notify.test.js`
- `test/init-uninstall.test.js`

The redundant popover changes in `DashboardViewModel.swift` and `test/macos-popover-anchor-window.test.js` have been dropped; both now match `main` exactly.

Validation on the rebased commit:

- `node --test test/codex-config-multiline-notify.test.js test/init-uninstall.test.js` — 55/55 passed
- `npm run ci:local` — 1557/1557 tests passed, plus copy/UI/architecture validations and the dashboard production build
- `git diff --check origin/main...HEAD` — passed

Ready for re-review.

File: test/init-uninstall.test.js [hunk 1]
```diff
@@ -271,6 +271,446 @@ test("notify handler avoids duplicating existing payload args when chaining", as
...
+          });
+          assert.equal(await waitForFile(markerPath, { timeoutMs: 500 }), null);
+
+          await runGeneratedNotifyHandler({
+            trackerDir: path.join(tmp, "tracker-bun-run-inline-cwd-rejects-process-cwd-script"),
+            notify: [fakeRuntimePath, "run", `--cwd=${bunCwdDir}`, "./process-cwd-only-bun-notify.js", "turn-ended"],
+          });
+          assert.equal(await waitForFile(markerPath, { timeoutMs: 500 }), null);
+        } finally {
+          await fs.rm(processCwdOnlyScript, { force: true });
```
[the remark this answers is not in the document]
Comment (lamther) [reply test/init-uninstall.test.js]: Fixed in `918b206`. The process-cwd-only fixture now derives a unique script basename from the per-test `mkdtemp` directory and uses that name in both Bun `--cwd` cases, while retaining cleanup in `finally`. This avoids overwriting a fixed repository file and prevents parallel-test collisions.

Validation: targeted regression 1/1, notify/config suite 55/55, and `npm run ci:local` 1557/1557 plus validations and dashboard build.

**Comment by lamther** [comment]
Follow-up: CodeRabbit's fixed-filename test concern was valid and is fixed in the final commit `918b206`. The fixture now derives a unique filename from its per-test temporary directory; production behavior and the four-file notify-only scope are unchanged.

I re-ran the targeted regression (1/1), the notify/config suite (55/55), and `npm run ci:local` (1557/1557 plus validations and dashboard build). The remote PR is now 7/7 checks green with no conflicts, and the review thread is marked outdated/resolved.

**Review by xiufengsun — APPROVED** [review]
Re-reviewed the final notify-only head. The earlier TOML, env/cwd validation, and redundant popover reload blockers are resolved; the popover scope was correctly dropped after rebasing. Targeted validation and all required checks pass. Approved for merge.
